### DISCUSSION TITLE
Loading sign text from Anvil world

### DISCUSSION THREAD

**Discussion by NotThatRqd** [Minestom/Minestom#1929] [Help] [answered]
There is very little documentation on Block Handlers, I was wondering how I would set one up so that signs would be loaded with their text when using an Anvil world. I think I'll have to be reading the text from the world file using NBTReader, which I think I could probably figure out. Mainly I'm wondering about setting up handlers for when a block is loaded in and stuff like that.

**Comment by TropicalShadow** [comment, marked as answer]
these are two solutions that work

sorry for them being Kotlin, but they should give a rough idea of what to do

Minestom normal
```kt
import net.kyori.adventure.key.Key
import net.minestom.server.instance.block.BlockHandler
import net.minestom.server.tag.Tag
import net.minestom.server.utils.NamespaceID

object SignHandler : BlockHandler {
    override fun getNamespaceId(): NamespaceID = NamespaceID.from(Key.key("minecraft:sign"))
    override fun getBlockEntityTags(): MutableCollection<Tag<*>> {
        return mutableListOf(
            Tag.Byte("GlowingText"),
            Tag.String("Color"),
            Tag.String("Text1"),
            Tag.String("Text2"),
            Tag.String("Text3"),
            Tag.String("Text4"),
        )
    }
}
```

Minestom CE (some people nowadays use a fork which has 1.20.1 support, signs changed for that versions so I made this

```kt

import net.kyori.adventure.key.Key
import net.minestom.server.instance.block.BlockHandler
import net.minestom.server.tag.Tag
import net.minestom.server.utils.NamespaceID

object SignHandler : BlockHandler {
    override fun getNamespaceId(): NamespaceID = NamespaceID.from(Key.key("minecraft:sign"))
    override fun getBlockEntityTags(): MutableCollection<Tag<*>> {
        return mutableListOf(
            Tag.Byte("is_waxed"),
            Tag.NBT("back_text"),
            Tag.NBT("front_text")
        )
    }
}
```

Also note you will need to register these somewhere

```kt
val bm = MinecraftServer.getBlockManager()

bm.registerHandler("minecraft:sign") { SignHandler }

        Block.values().forEach {
            if (it.name().endsWith("sign")) {
                bm.registerHandler(it.namespace()) { SignHandler }
            }
        }
```

**Reply by NotThatRqd** [reply]
Thanks! I'm actually using Kotlin and I got it working, but one other question: 

What actually is `getBlockEntityTags`? Do you know of a place to find info on what Entity Tags and what order they go in for each block?

**Reply by TropicalShadow** [reply]
Glad the code works,
I haven't found a perfect way of monitoring the tags but using minecraft's fandom page you can see their tags.

i.e. https://minecraft.fandom.com/wiki/Sign#Block_data

**Reply by NotThatRqd** [reply]
Oh I see Block Entity is another name for tile entity!

Tysm man :)
